### PR TITLE
Better logging for "cannot fuse" reasons

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114917
* #115019
* __->__ #115003

This was invaluable when I was debugging #114917. Without the node names
in the log message, it was difficult to make sense of them.

However, I did not want to bloat the number of LOC with this change.
Thus, instead of calling `debug()` directly with the node arguments, I
made a new callable class WhyNoFuse to partially apply the node
arguments at the top of each fusion-checking method. WhyNoFuse generates
the logging string only when its `__str__` method gets called, so there
is minimal overhead when logging is disabled.

I also removed the various logging 'tags' like "vert:1" / "triton:1" --
the log messages themselves are unique enough that the user can identify
them without the tag.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -41,6 +52,27 @@
 fusion_log = torch._logging.getArtifactLogger(__name__, "fusion")
 
 
+class WhyNoFuse:
```
Comment (Skylion007) [inline torch/_inductor/scheduler.py:55]: Any reason why this shouldn't just be a dataclass? With slots=True?

Comment (int3) [reply torch/_inductor/scheduler.py:55]: that was my initial attempt :) but I realized that `slots=True` is only supported in Python >= 3.10

Comment (Skylion007) [reply torch/_inductor/scheduler.py:55]: Hmm, so apparently you can still use a dataclass, but define slots yourself as long as you don't use default args in them. But yeah maybe just add a comment about why so we can fix it when we drop support for 3.9

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**